thavalon-server/src/database/games/db_game.rs

```diff
@@ -0,0 +1,282 @@
+//! Game collection related functions and structs
+use crate::database::get_database;
+
+use std::{collections::HashSet, iter};
+
+use async_trait::async_trait;
+use chrono::Utc;
+use mongodb::{
+    bson::{self, doc, oid::ObjectId, Document},
+    error::Error,
+    options::{InsertOneOptions, ReplaceOptions, UpdateModifications, UpdateOptions},
+    results::{InsertOneResult, UpdateResult},
+    Collection,
+};
+use rand::{distributions::Alphanumeric, Rng};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+const GAME_COLLECTION: &str = "thavalon_games";
+const FRIEND_CODE_LENGTH: usize = 4;
+
+/// Contains errors related to database games.
+#[derive(PartialEq, Error, Debug)]
+pub enum DBGameError {
+    #[error("The game could not be created.")]
+    CreationError,
+    #[error("An error occurred while updating the game in the database.")]
+    UpdateError,
+    #[error("Invalid state for the requested update.")]
+    InvalidStateError,
+}
+
+/// Enum representing the three possible states of a game in the DB.
+#[derive(PartialEq, Serialize, Deserialize, Debug, Clone)]
+pub enum DBGameStatus {
+    Lobby,
+    InProgress,
+    Finished,
+}
+
+/// Struct representing a single database game.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct DatabaseGame {
+    _id: ObjectId,
+    friend_code: String,
+    players: HashSet<String>,
+    display_names: HashSet<String>,
+    status: DBGameStatus,
+    created_time: i64,
+    start_time: Option<i64>,
+    end_time: Option<i64>,
+    snapshot_id: Option<String>,
+}
+
+impl DatabaseGame {
+    /// Creates a new DB game entry and returns a DatabaseGame
+    ///
+    /// # Returns
+    ///
+    /// * `DatabaseGame` on success. `GameError::CreationError` on failure.
+    pub async fn new() -> Result<Self, DBGameError> {
+        log::info!("Creating a new database game.");
+        let collection = get_database().await.collection(GAME_COLLECTION);
+        let _id: ObjectId = match collection.insert_one(doc! {}, None).await {
+            Ok(result) => {
+                bson::from_bson(result.inserted_id).expect("Could not deserialize new game _id.")
+            }
+            Err(e) => {
+                log::error!("ERROR: failed to create new game. {}.", e);
+                return Err(DBGameError::CreationError);
+            }
+        };
+
+        let friend_code;
+        {
+            let mut rng = rand::thread_rng();
+            friend_code = iter::repeat(())
+                .map(|()| rng.sample(Alphanumeric))
+                .take(4)
+                .collect::<String>()
+                .to_uppercase();
```

Comment:
We should probably have a helper to generate a random string somewhere, it looks like this code is repeated a couple times